Species homepage stats

msmanoj opened the pull request:
## Type

- [ ] Bug fix
- [x] New feature
- [ ] Improvement to existing feature

## Related JIRA Issue(s)
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-758

## Description
- Created the shared component `ExpandableSection`
- Using a local sample data file which is populated using the following sheet:
https://docs.google.com/spreadsheets/d/1NYANlVmB4tzw4cIluOI4kUWtXRd1P9jU-CNbYOCqSzA/edit#gid=0

## Deployment URL
http://species-homepage-stats.review.ensembl.org/species/homo_sapiens_GCA_000001405_28

## Views affected
Species homepage

### Other effects
<!--
_List any other functionality that may be affected or which requires additional changes, such as saved configurations. Please add an explanation if, for example, no test is needed._
-->

- [ ] I have checked any requirements for Google Analytics
- [x] I have created any required tests
- [ ] The PR has as its final commit a WASM/JS update commit if any Rust files were updated.

msmanoj commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainView.scss:
```diff
@@ -6,5 +8,124 @@
+}
+
+// TODO: Remove Me
+.statsWrapper { 
```
TODO: This needs to be updated once the PR https://github.com/Ensembl/ensembl-client/pull/351 is merged

imransl commented on src/ensembl/src/shared/components/expandable-section/ExpandableSection.tsx:
```diff
@@ -0,0 +1,92 @@
+        <Chevron className={chevronClasses} />
+      </div>
+
+      {isExpanded && (
+        <div className={expandedContentClassNames}>{props.expandedContent}</div>
+      )}
+
+      {!isExpanded && (
+        <div className={collapsedContentClassNames}>
+          {props.collapsedContent}
+        </div>
+      )}
```
How about using the ternary operator for this?

msmanoj replied:
That's done

azangru commented on src/ensembl/src/content/app/species/SpeciesPage.tsx:
```diff
@@ -32,30 +31,21 @@ import SpeciesAppBar from './components/species-app-bar/SpeciesAppBar';
   const dispatch = useDispatch();
 
   useEffect(() => {
     dispatch(setActiveGenomeId(genomeId));
+    dispatch(fetchGenomeData(genomeId));
```
![image](https://user-images.githubusercontent.com/6834224/95137330-b717a600-075f-11eb-8118-329cf223e235.png)

[link](https://react-redux.js.org/api/batch)

Apparently, it only makes sense from within something like thunks, not from inside React components.

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainView.scss:
```diff
@@ -1,3 +1,5 @@
+@import 'src/styles/common';
```
Lots of 3 (or 4?) spaces used as indents in this file. Don't we normally use two spaces?

msmanoj replied:
That's done

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainView.scss:
```diff
@@ -6,5 +8,120 @@
+  .stats{
+      grid-column: stats;
+  }
+}
```
Also, our editors will automatically add an empty line in the end of a file. What happened to your VSCode? 😄

msmanoj replied:
I've added line break but I'm not sure what's causing it. I'll keep an eye on it.

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,217 @@
+  StatsSection,
+  sectionGroupsMap
+} from '../../state/general/speciesGeneralHelper';
+import { ExampleFocusObject } from 'src/shared/state/genome/genomeTypes';
```
order of imports

msmanoj replied:
I always get confused with this :(

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,217 @@
+              anchor={anchorRef.current as HTMLDivElement}
+              onOutsideClick={() => setShowPointerBox(false)}
+              position={Position.BOTTOM_RIGHT}
+              container={containerRef.current}
```
The point of the container prop is to make the pointer box try to avoid extending beyond the borders of a given outer element:

![image](https://user-images.githubusercontent.com/6834224/95140737-7a4fad00-0767-11eb-84e1-3db8d43c7f89.png)

It defaults to `window` and is useless here.

msmanoj replied:
Yes, that's right.

Any idea how the following can be prevented? Should it be closed as soon as the user starts to scroll?
![Kapture 2020-10-06 at 14 21 59](https://user-images.githubusercontent.com/3085815/95207335-aa3a9700-07df-11eb-8776-6e4598050823.gif)

azangru replied:
> Should it be closed as soon as the user starts to scroll?

That's what the tooltips do. I'd say yes.

One other option is to try the `renderInsideAnchor` prop; then the pointer box will be placed into the "Example gene" span rather than at the top level of the window. This might cause other layout problems — or it may work out fine :-)

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,217 @@
+        >
+          {exampleLinkText}
+          {showPointerBox && (
+            <PointerBox
```
This popup is recurring theme with Andrea; there was time when she had it on the home page. It looks very much like a separate component.

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,217 @@
+          {exampleLinkText}
+          {showPointerBox && (
+            <PointerBox
+              anchor={anchorRef.current as HTMLDivElement}
```
The type above says that it's `HTMLSpanElement`. But probably better just to add `anchorRef.current` to the condition above (`showPointerBox && anchorRef.current && ...`).

msmanoj replied:
That's done

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,217 @@
+
+  if (!exampleLinks) {
+    return null;
+  }
```
❓ 
why would ExampleLinks be a valid component without the example links?

msmanoj replied:
Removed the check and made the prop mandatory

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,217 @@
+const ExampleLinks = (props: {
+  exampleLinks?: Partial<urlObj>;
+  exampleLinkText?: string;
+}) => {
```
Several considerations:
- The component appears as a single link, although it might show the popup with buttons (compare our other ExampleLinks components, where we are showing multiple links). Perhaps better call it `ExampleFeatureLink` or `ViewExample`? Is it possible that this is a reusable component? It doesn't look particularly specific to `SpeciesMainViewStats`.
- The `example` part of the `exampleLinks` prop is redundant. `links` should be sufficient, as it is called in the `ViewInApp` component.
- The `exampleLinkText` can then simply become `children`
- When there is only one app in which the example can be viewed (e.g. example region can only be viewed in GenomeBrowser), I am sure there is no point in showing the popup. We should just redirect the user there.

msmanoj replied:
> Is it possible that this is a reusable component? It doesn't look particularly specific to SpeciesMainViewStats.

How about creating something like `ViewInAppPopup`?

azangru replied:
Sounds good.

azangru commented on src/ensembl/src/content/app/species/state/general/speciesGeneralSlice.ts:
```diff
@@ -13,15 +13,71 @@
+  );
+
+  const speciesStats = [
+    ...Object.values(SpeciesStatsSections).map((section) =>
```
why the spread?

msmanoj replied:
Removed it

azangru commented on src/ensembl/src/content/app/species/state/general/speciesGeneralHelper.ts:
```diff
@@ -0,0 +1,419 @@
+import { buildFocusIdForUrl } from 'src/shared/state/ens-object/ensObjectHelpers';
+import { urlObj } from 'src/shared/components/view-in-app/ViewInApp';
+
+export enum SpeciesStatsSections {
```
1) Enums are usually named in singular, because they represent not a set of things, but one thing belonging to a set. See [example in typescript docs](https://www.typescriptlang.org/docs/handbook/enums.html#numeric-enums).

2) Aren't we using enums a bit too much? What would this code look like if we didn't use enums?

msmanoj replied:
> Aren't we using enums a bit too much? What would this code look like if we didn't use enums?
Any suggestions for alternatives?

azangru commented on src/ensembl/src/content/app/species/state/general/speciesGeneralHelper.ts:
```diff
@@ -0,0 +1,419 @@
+  [SpeciesStatsSections.CODING_STATS]: {
+    title: Stats.CODING_GENES,
+    hasExampleLink: true,
+    exampleLinkText: 'Example gene',
```
`hasExampleLink` and `exampleLinkText` seem always to coincide. `hasExampleLink` looks like a redundant property; let's get rid of it

azangru replied:
Yep.

msmanoj replied:
That's done

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,217 @@
+
+const SpeciesMainViewStats = (props: Props) => {
+  useEffect(() => {
+    if (!props.genomeStats && props.exampleFocusObjects?.length) {
```
Ouch! This is a very confusing dependency. Fetching genome stats should not depend on whether or not we have fetched example focus objects.

msmanoj replied:
That's right. Removed `props.exampleFocusObjects?.length` from here and moved it to the if statement below.

msmanoj replied:
Example links are getting generated while the stats are getting processed. This is why I'm trying to make sure the exampleObjects are available before fetching the genomeStats.

azangru commented on src/ensembl/src/content/app/species/components/species-stats/SpeciesStats.scss:
```diff
@@ -3,7 +3,8 @@
 
 .wrapper {
-  width: 250px;
+  min-width: 250px;
```
why min-width rather than width?

msmanoj replied:
Missed to clean up

azangru commented on src/ensembl/src/content/app/species/components/species-stats/SpeciesStats.scss:
```diff
@@ -3,7 +3,8 @@
 .wrapper {
-  width: 250px;
+  min-width: 250px;
+  margin-bottom:30px;
```
space after colon

msmanoj replied:
Update it

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,217 @@
+                );
+              })}
+            </div>
+            {group_index === 0 && (
```
I'd argue the way it's written now (`group_index === 0`) is ever so slightly more readable. With `!group_index` in this context you would have to mentally translate it back: "why is the group_index falsey? oh, that's because it's zero". The way it's written now saves this mental step.

Probably both ways of writing it are equally fine.

imransl replied:
If so then shouldn't we be using the same method in a single file as line 140 uses the other method

azangru replied:
🤷‍♂️

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,217 @@
+    <div className={styles.expandedContent}>
+      {groups.map((group, group_index: number) => {
+        const { title, stats } = group;
+        const primaryStatsClassName = classNames(styles.primaryStats, {
```
Am I reading this correctly and the calculation of this class name does not depend on anything happening inside the loop? If so, then it should be calculated outside the `.map`

msmanoj replied:
That's done

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainView.scss:
```diff
@@ -6,5 +8,120 @@
+  padding: 13px 20px 0;
+  display: grid;
+  grid-template-columns: [title] 120px [stats] auto [example-link] 150px;
+  align-items: flex-start;
```
the value of align-items for `display: grid` does not use the prefix `flex-`

msmanoj replied:
Removed `align-items` as it's not required

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainView.scss:
```diff
@@ -6,5 +8,120 @@
+  padding: 13px 20px 0;
+  display: grid;
+  grid-template-columns: [title] 120px [stats] auto;
+  align-items: flex-start;
```
The value of align-items for display: grid does not use the prefix flex-

Also, there is a lot of repetition between the `primaryStats` and `secondaryStats` classes. Won't a single class with some modifiers be better?

msmanoj replied:
>The value of align-items for display: grid does not use the prefix flex-

Removed it

>Also, there is a lot of repetition between the primaryStats and secondaryStats classes. Won't a single class with some modifiers be better?

That's done

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainView.scss:
```diff
@@ -6,5 +8,120 @@
+      }
+
+      .secondaryUnit{
+          color: #6f8190;
```
this color is defined in our design system as a variable

msmanoj replied:
Updated it

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,217 @@
+            : true
+        });
+        const primaryOrSecondaryClassName =
+          group_index === 0 ? primaryStatsClassName : styles.secondaryStats;
```
I tried to figure this out from the code and from the way the page looks, and couldn't. What's the difference between "primary stats" and "secondary stats"? Is it just the first stats group versus all others? Then shouldn't it be something like `statsGroup` vs `firstStatsGroup`? Or, for more convenience with SCSS, expressing it in the BEM style: `statsGroup` vs `statsGroup_first`?

msmanoj replied:
Updated to use `first-of-type` and removed the line above.

azangru commented on src/ensembl/src/content/app/species/state/general/speciesGeneralSelectors.ts:
```diff
@@ -18,3 +18,13 @@ import { RootState } from 'src/store';
   state.speciesPage.general.activeGenomeId;
+
+export const getActiveGenomeStats = (state: RootState) => {
+  const activeGeonmeId = getActiveGenomeId(state);
```
Geonme – 🙅‍♂️

msmanoj replied:
That's done

azangru commented on src/ensembl/src/shared/components/expandable-section/ExpandableSection.scss:
```diff
@@ -0,0 +1,39 @@
+      transform: rotate(-90deg);
+      
+    }
+  }
```
Also suddenly 4-space indents and no empty line in the end of the file 😞

msmanoj replied:
Fixed it

azangru commented on src/ensembl/stories/shared-components/expandable-section/ExpandableSection.stories.scss:
```diff
@@ -0,0 +1,87 @@
+    .stats{
+        grid-column: stats;
+    }
+}
```
again, 4-space indents and no trailing empty line 😞

msmanoj replied:
Fixed it

azangru commented on src/ensembl/src/content/app/species/sample-data.ts:
```diff
@@ -14,30 +14,605 @@
+      Chromosomes: '25',
+      'Toplevel sequences': '640',
+      'Component sequences': '36734',
+      '% GC': '38.87'
```
The structure presented here is very UI-specific, where the key in the json object is the same string as the label in the UI. I strongly feel that the proper long-term approach is to design a data structure that is likely to come from the backend, then figure out how to transform this data structure into something consumable by the frontend, and finally pass the proposed data structure to the backend for implementation. I am almost positive the the backend will not want to format the keys of the JSON this way.

A conversation with the backend team might help decide on the best shape of the json. @ens-st3, @jgtate what is your advise?

azangru commented on src/ensembl/src/content/app/species/state/general/speciesGeneralSlice.ts:
```diff
@@ -13,15 +13,71 @@
+  getStatsForSection,
+  StatsSection,
+  SpeciesStatsSections
+} from 'src/content/app/species/state/general/speciesGeneralHelper';
```
order of imports

msmanoj replied:
That's done

azangru commented on src/ensembl/src/shared/components/expandable-section/ExpandableSection.test.tsx:
```diff
@@ -0,0 +1,95 @@
+    expect(() => wrapper).not.toThrow();
+  });
+
+  it('hides expandedContent when isExpanded is false', () => {
```
`it('shows only collapsed content when collapsed'...)`

msmanoj replied:
That's done

msmanoj commented on src/ensembl/src/content/app/species/state/general/speciesGeneralHelper.ts:
```diff
@@ -0,0 +1,419 @@
+import { SpeciesStatsProps } from 'src/content/app/species/components/species-stats/SpeciesStats';
+import { ExampleFocusObject } from 'src/shared/state/genome/genomeTypes';
+
+import { sampleData } from '../../sample-data';
```
I thought relative paths up to two levels are okay?

imransl replied:
I can't remember if we had a consensus on this. If we did, then let's just leave it as it is.

azangru commented on src/ensembl/src/shared/components/expandable-section/ExpandableSection.test.tsx:
```diff
@@ -0,0 +1,95 @@
+    expect(wrapper.find('.collapsedContent')).toHaveLength(0);
+  });
+
+  it('calls onToggle with expanded status when the chevron wrapper is clicked', () => {
```
`it('calls onToggle prop when collapsing or expanding'...)`

and then the test should check that onToggle was called with `false` if the section was expanded and with `true` if it was collapsed.

msmanoj replied:
That's done

msmanoj commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,217 @@
+  const { exampleLinkText } = sectionGroupsMap[section];
+  return (
+    <div className={styles.expandedContent}>
+      {groups.map((group, group_index: number) => {
```
That's done

msmanoj commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,217 @@
+          <div key={group_index} className={primaryOrSecondaryClassName}>
+            <span className={styles.title}>{title}</span>
+            <div className={styles.stats}>
+              {stats.map((groupStats, group_index: number) => {
```
That's done

msmanoj commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,217 @@
+              {stats.map((groupStats, group_index: number) => {
+                return (
+                  <div key={group_index}>
+                    {groupStats.map((stat, stat_index: number) => {
```
That's done

msmanoj commented on src/ensembl/src/content/app/species/state/general/speciesGeneralHelper.ts:
```diff
@@ -0,0 +1,419 @@
+          genomeId: genome_id,
+          entityId: focusId
+        })
+      : undefined;
```
That's done

azangru commented on src/ensembl/src/shared/components/view-in-app-popup/ViewInAppPopup.scss:
```diff
@@ -0,0 +1,17 @@
+  
+  .pointerBoxPointer {
+    fill: $black;
+  }
```
Indents and terminal empty line again 😞

azangru commented on src/ensembl/stories/shared-components/expandable-section/ExpandableSection.stories.scss:
```diff
@@ -0,0 +1,85 @@
+  padding: 13px 20px;
+  display: grid;
+  grid-template-columns: [title] 120px [stats] auto [example-link] 200px;
+  align-items: flex-start;
```
here and below — no flex

azangru commented on src/ensembl/stories/shared-components/expandable-section/ExpandableSection.stories.scss:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    .primaryUnit {
+      color: #6f8190;
```
here and below — use scss variable

azangru commented on src/ensembl/src/content/app/species/sample-data.ts:
```diff
@@ -14,30 +14,615 @@
+type RawSpeciesStats = {
+  [genomeId: string]: {
+    [key in SpeciesStatsSection]: {
+      [key: string]: string;
```
This is something I would definitely urge we change. The type of field value can be at least one of the following: string, number, null

azangru commented on src/ensembl/src/content/app/species/sample-data.ts:
```diff
@@ -14,30 +14,615 @@
+  homo_sapiens_GCA_000001405_28: {
+    assembly_stats: {
+      scientific_name: 'Homo sapiens',
+      sex: 'NA',
```
All the NAs should become nulls

azangru commented on src/ensembl/src/content/app/species/sample-data.ts:
```diff
@@ -14,30 +14,615 @@
+      assembly_name: 'GRCh38.p13',
+      assembly_accession: 'GCA_000001405.28',
+      assembly_date: '2019-02-28',
+      contig_n50: '56413054',
```
All the numbers should be of type number.

azangru commented on src/ensembl/src/content/app/species/sample-data.ts:
```diff
@@ -14,30 +14,615 @@
+    assembly_stats: {
+      scientific_name: 'Homo sapiens',
+      sex: 'NA',
+      breed_cultivar_strain: 'NA',
```
This field can probably just be called "strain". Whichever is the preferred term for this taxonomic group at Ensembl.

azangru commented on src/ensembl/src/content/app/species/sample-data.ts:
```diff
@@ -14,30 +14,615 @@
+      average_sequence_length: '3443.15',
+      average_CDS_length: '1139.02',
+      shortest_gene: '8',
+      longest_gene: '2473539',
```
shortest_gene_**length**? longest_gene_**length**?

msmanoj replied:
That's done

azangru commented on src/ensembl/src/content/app/species/sample-data.ts:
```diff
@@ -14,30 +14,615 @@
+      coding_genes: '20420',
+      average_genomic_span: '66885.16',
+      average_sequence_length: '3443.15',
+      average_CDS_length: '1139.02',
```
CDS -> cds (lower-case)

azangru commented on src/ensembl/src/content/app/species/state/general/speciesGeneralHelper.ts:
```diff
@@ -0,0 +1,462 @@
+  const formattingOptions = statsFormattingOptions[props.primaryKey];
+
+  if (!isNaN(Number(primaryValue))) {
+    primaryValue = getCommaSeparatedNumber(Number(primaryValue));
```
Why don't we expect numbers to be of type `number`? Why are they string/number?

_(And, consequently, why isn't the if-condition `if(typeof primaryValue === 'number)`?)_

msmanoj replied:
That's all updated now

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainViewStats.tsx:
```diff
@@ -0,0 +1,173 @@
+
+type Props = {
+  activeGenomeId: string | null;
+  genomeStats: GenomeStats | undefined;
```
Why would it ever be undefined? Does it make sense for a component responsible for displaying stats to receive an undefined instead of stats?

azangru commented on src/ensembl/src/content/app/species/state/general/speciesGeneralHelper.ts:
```diff
@@ -0,0 +1,462 @@
+  section: SpeciesStatsSection;
+  exampleLinks?: Partial<urlObj>;
+  primaryStats?: IndividualStat;
+  secondaryStats?: IndividualStat;
```
What do primary stats and secondary stats mean for a stats section?

msmanoj replied:
They are the two stats that gets displayed in the header when the section is collapsed. Would `firstHeaderStat` & `secondHeaderStat` work?

azangru replied:
What if we call them something like `summaryStats` or `briefStats` and rename `groups` to `fullStats`?

Then `primaryStats` and `secondaryStats` will become a single property, which is of type `IndividualStat[]`.

msmanoj replied:
That's done

azangru commented on src/ensembl/src/content/app/species/state/general/speciesGeneralHelper.ts:
```diff
@@ -0,0 +1,462 @@
+    const focusId = regionExample?.id
+      ? buildFocusIdForUrl({
+          type: 'region',
+          objectId: regionExample?.id
```
at this point, `regionExample` definitely exists (same above with `geneExample`).

msmanoj replied:
Updated it

azangru commented on src/ensembl/src/content/app/species/state/general/speciesGeneralHelper.ts:
```diff
@@ -0,0 +1,462 @@
+    label: 'Total coding sequence length'
+  },
+  [Stats.AVERAGE_GC_CONTENT]: {
+    primaryValuePostfix: '%',
```
what's the difference between `primaryUnit` and `primaryValuePostfix`?

msmanoj replied:
I've removed the need to `primaryValuePostfix` as it's simpler to append the postfix value in the helper function itself.

azangru commented on src/ensembl/src/content/app/species/components/species-stats/SpeciesStats.tsx:
```diff
@@ -81,12 +83,17 @@ const SpeciesStats = (props: SpeciesStatsProps) => {
-        <span className={styles.primaryValue}>{props.primaryValue}</span>
+        <span className={styles.primaryValue}>
+          {props.primaryValue}
+          {props.primaryValuePostfix || null}
```
no need for `|| null`

msmanoj replied:
Removed the prop `primaryValuePostfix` as mentioned in https://github.com/Ensembl/ensembl-client/pull/352#discussion_r503607775

azangru commented on src/ensembl/src/content/app/species/state/general/speciesGeneralHelper.ts:
```diff
@@ -0,0 +1,462 @@
+
+  if (!primaryValue) {
+    return;
+  }
```
Erm, can primaryValue be `0`?

msmanoj replied:
Removed it as this block is no longer required.

azangru commented on src/ensembl/src/content/app/species/state/general/speciesGeneralHelper.ts:
```diff
@@ -0,0 +1,462 @@
+    label: formattingOptions?.label || props.primaryKey,
+    primaryValue,
+    primaryUnit: formattingOptions?.primaryUnit,
+    primaryValuePostfix: formattingOptions?.primaryValuePostfix
```
If part of the formatting (`getCommaSeparatedNumber`) is happening in this function, why not add primaryValuePostfix to the comma-separated number here, without passing it to the component?

msmanoj replied:
That's done

azangru commented on src/ensembl/src/content/app/species/sample-data.ts:
```diff
@@ -14,30 +14,615 @@
+      taxonomy_id: 9606,
+      assembly_name: 'GRCh38.p13',
+      assembly_accession: 'GCA_000001405.28',
+      assembly_date: '2019-02-28',
```
Do you want to logically separate the fields that describe the assembly (scientific_name, sex, taxonomy_id, assembly_name, assembly_accession, assembly_date) into a dedicated section (`assembly_description`?, `assembly_summary`?)? It feels like a perceptibly different kind of data.

msmanoj replied:
That sounds like a good idea. `assembly_summary` feels like a better name.

msmanoj replied:
I tried to make this change but currently these sections each represents an individual `ExpandableComponent`. Creating a separate section called `assembly_summary` which will not have it's own `ExpandableComponent` feels like it will cause more confusion. Would it be worth getting back to this later on?

azangru replied:
> Would it be worth getting back to this later on?

I think it would be nice to have this separation by the time we give our json back to Fergal so that he can update it programmatically. Were you thinking of getting back to it earlier than that (e.g. during your work on the sidebar)?

msmanoj replied:
I've tried to do it in this PR itself. Please check if it looks okay.

azangru commented on src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/DefaultTranscriptsList.scss:
```diff
@@ -60,6 +60,7 @@
   grid-column: 3/4;
   position: relative;
   top: 2px;
+  cursor: pointer;
```
I'm surprised. How home you have something for entity viewer in this PR? 😄

msmanoj replied:
I once merged the layout update branch before it got merged into dev. That could be why this got left out. I think I need to remove this since it is shown as a difference.

msmanoj replied:
That's done

azangru commented on src/ensembl/src/content/app/species/state/general/speciesGeneralHelper.ts:
```diff
@@ -0,0 +1,477 @@
+): IndividualStat | undefined => {
+  let primaryValue = props.primaryValue;
+
+  const { primaryValuePostfix, headerUnit } = statsFormattingOptions[
```
if `StatsFormattingOptions` type really has a field `{primaryValuePostfix?: string;}`, then you can use the default value for primaryValuePostfix when destructuring (`const { primaryValuePostfix = '' , ... } = ...`). Then you wouldn't need to apply a ternary operator to check that primaryValuePostfix exists below.

msmanoj replied:
That's done

azangru commented on src/ensembl/src/content/app/species/sample-data.ts:
```diff
@@ -14,30 +14,637 @@
+      gc_percentage: 38.87
+    },
+    coding_stats: {
+      scientific_name: 'Homo sapiens',
```
This has been moved to assembly summary. Same for other sections and other species below.

msmanoj replied:
Remove it

azangru replied:
I'm still seeing it in multiple places in the file (https://github.com/Ensembl/ensembl-client/blob/ac354582bcc65b250a2e298b70d189246e16669f/src/ensembl/src/content/app/species/sample-data.ts).

msmanoj replied:
Should be gone now?

azangru commented:
As a reminder from code review:

![image](https://user-images.githubusercontent.com/6834224/95985981-391d6400-0e1d-11eb-9306-60001cd04e04.png)

- stuff that's currently in assembly_summary should not appear in the assembly stats section
- need to re-check groupings of data in sections with the XD (we noticed that the coding genes section does not contain all stats that are in the XD)

msmanoj commented:
> As a reminder from code review:
> 
> ![image](https://user-images.githubusercontent.com/6834224/95985981-391d6400-0e1d-11eb-9306-60001cd04e04.png)
> 
> * stuff that's currently in assembly_summary should not appear in the assembly stats section
> * need to re-check groupings of data in sections with the XD (we noticed that the coding genes section does not contain all stats that are in the XD)

That's all done

azangru commented:
@msmanoj something isn't right:

- pseudogenes among coding genes:
![image](https://user-images.githubusercontent.com/6834224/96017755-46007e80-0e42-11eb-93b4-bdc7d05243ce.png)

- pseudogenes in Analysis:
![image](https://user-images.githubusercontent.com/6834224/96017834-5fa1c600-0e42-11eb-8905-176a15999238.png)

- pseudogenes in Transcripts:
![image](https://user-images.githubusercontent.com/6834224/96017935-7fd18500-0e42-11eb-9bd7-5b1847d34337.png)

- pseudogenes everywhere...
![image](https://user-images.githubusercontent.com/6834224/96017994-9677dc00-0e42-11eb-9b89-42cef36b7920.png)

msmanoj commented:
> @msmanoj something isn't right:
> 
> * pseudogenes among coding genes:
>   ![image](https://user-images.githubusercontent.com/6834224/96017755-46007e80-0e42-11eb-93b4-bdc7d05243ce.png)
> * pseudogenes in Analysis:
>   ![image](https://user-images.githubusercontent.com/6834224/96017834-5fa1c600-0e42-11eb-8905-176a15999238.png)
> * pseudogenes in Transcripts:
>   ![image](https://user-images.githubusercontent.com/6834224/96017935-7fd18500-0e42-11eb-9bd7-5b1847d34337.png)
> * pseudogenes everywhere...
>   ![image](https://user-images.githubusercontent.com/6834224/96017994-9677dc00-0e42-11eb-9b89-42cef36b7920.png)

That should now be fixed. It was because of keys getting overridden in the `statsFormattingOptions`

azangru commented:
Coding genes, analysis

Design:
![image](https://user-images.githubusercontent.com/6834224/96046039-c044f900-0e6a-11eb-8ed0-5d68717d3fc4.png)

Deployment:
![image](https://user-images.githubusercontent.com/6834224/96046063-ccc95180-0e6a-11eb-9c8f-8318eab5ea3b.png)

msmanoj commented:
@azangru - the missing `Average CDS length` is now back. I think `Average coding exons per transcript` is the duplicate of the one that's displayed in the `Transcripts` group. I will check this with Andrea and get back.

I've increase the PrimaryValue font size to 25px and it seems to be matching the XD.

azangru commented:
Notice that font size in the collapsed and expanded section are the same in the design:

![image](https://user-images.githubusercontent.com/6834224/96112065-f79ebe80-0ed9-11eb-910f-92e7a0ef0287.png)

But different in the deployment:

![image](https://user-images.githubusercontent.com/6834224/96112099-05ecda80-0eda-11eb-952a-1d468df37813.png)

> I've increased the PrimaryValue font size to 25px and it seems to be matching the XD.

Probably should ask Andrea about font sizes to make sure?

azangru commented on src/ensembl/src/content/app/species/components/species-main-view/SpeciesMainView.scss:
```diff
@@ -13,7 +13,8 @@
   padding: 5px 20px 40px 120px;
-  max-width: 1360px;
+  max-width: 1381px;
+  font-size: 13px;
```
This is the default font size (see _settings.scss). I tried to disable this rule in the deployment, and didn't see any change. It's probably redundant.

msmanoj replied:
Removed it

azangru commented on src/ensembl/src/content/app/species/components/species-stats/SpeciesStats.scss:
```diff
@@ -7,38 +7,31 @@ $secondary-text-color: $dark-grey;
   margin-bottom: 30px;
   display: inline-block;
   vertical-align: top;
+  font-size: 13px;
```
same here: default font size, inherited from the `html` element, doesn't change anything if disabled, so likely redundant

msmanoj replied:
Removed it

azangru commented:
Non-coding genes, analysis, human:
![image](https://user-images.githubusercontent.com/6834224/96126070-ebbbf800-0eeb-11eb-8882-53195b147ae4.png)

should have `bp` units

![image](https://user-images.githubusercontent.com/6834224/96126536-10b06b00-0eec-11eb-82ee-420728474a6a.png)
genes -> gene, for both longest and shortest

![image](https://user-images.githubusercontent.com/6834224/96127483-52d9ac80-0eec-11eb-934e-a1e542d18902.png)
no `bp`

![image](https://user-images.githubusercontent.com/6834224/96127788-684ed680-0eec-11eb-8af7-8bc785856f40.png)
`bp` for both

![image](https://user-images.githubusercontent.com/6834224/96127907-861c3b80-0eec-11eb-9102-7ec1f7b323b7.png)
no `bp` for all

![image](https://user-images.githubusercontent.com/6834224/96127982-a3e9a080-0eec-11eb-85b6-172e74b679db.png)
Ah! we are getting wrapped labels! Check w/Andrea what to do

![image](https://user-images.githubusercontent.com/6834224/96128098-d5fb0280-0eec-11eb-922c-6598fdf08c18.png)
"Top level" is written as two words in XD, although I am sure it should be spelled with a hyphen (top-level). If no-one cares about the hyphen, spell in two words.

Only checked for human. Hopefully, it will be the same for other species.

msmanoj commented:
@azangru  - I have added the missing units and removed where they are not needed. Regarding the wrapping of labels, Andrea mentioned not to worry about them now as she is planning to review them all and also get some stats re-arranged.
We can get those changes done in a subsequent PR once the design is updated.

azangru commented:
I pushed a commit for the labels you missed in your last one.

Let's merge this thing in 😃
